src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java

```diff
@@ -453,32 +455,73 @@ private CompletableFuture<Response> checkFor407(ExchangeImpl<T> ex, Throwable t,
     // for the 100-Continue response
     private CompletableFuture<Response> expectContinue(ExchangeImpl<T> ex) {
         assert request.expectContinue();
+
+        long responseTimeout = 5;
+        if (request.timeout().isPresent() && request.timeout().get().getSeconds() < responseTimeout) {
+            responseTimeout = request.timeout().get().getSeconds();
+        }
+
         return ex.getResponseAsync(parentExecutor)
+                .orTimeout(responseTimeout, TimeUnit.SECONDS)
+                .handle((response, exception) -> {
+                    CompletableFuture<Response> cf;
+                    if (exception != null) {
+                        if (exception instanceof TimeoutException) {
+                            // Timeout has occurred, this SHOULD result in the request being sent anyway
+                            // return the null response to be handled below
+                            if (debug.on())
+                                debug.log("Request timeout waiting for response");
+                            cf = CompletableFuture.completedFuture(response);
+                        } else {
+                            // All other exceptions get returned as failedFuture
+                            cf = CompletableFuture.failedFuture(exception);
+                        }
+                        return cf;
+                    }
+                    // If there was no exception then return the response
+                    cf = CompletableFuture.completedFuture(response);
+                    return cf;
+                }).thenCompose(Function.identity())
```

Comment:
Lines 465-484 could be simplified to a single line call:
```
     .completeOnTimeout(null, responseTimeout, TimedUnit.XXXX)
```
since all we do in there is to replace the `TimeoutException` with a null response.